File: nodejs/src/state/CancelledState.ts
import OrderState, { IOrder } from "./OrderState";

class CancelledState extends OrderState {
    constructor(order: IOrder) { super(order); }

    getName(): string { return "CANCELLED"; }

    next(): void {
        console.log(`⚠️ Đơn #${this.order.id}: Đã huỷ rồi, không thể tiến thêm`);
    }

    cancel(): void {
        console.log(`⚠️ Đơn #${this.order.id}: Đã huỷ rồi rồi`);
    }
}

export default CancelledState;

File: spring/demo/src/main/java/com/example/demo/state/CancelledState.java
package com.example.demo.state;

import com.example.demo.model.Order;

public class CancelledState extends OrderState {
    public CancelledState(Order order) {
        super(order);
    }

    @Override
    public String getName() {
        return "CANCELLED";
    }

    @Override
    public void next() {
        System.out.println("Đơn #" + order.getId() + ": Đơn đã hủy, không thể chuyển tiếp");
    }

    @Override
    public void cancel() {
        System.out.println("Đơn #" + order.getId() + ": Đơn đã hủy trước đó");
    }
}
